src/chainstate/stacks/miner.rs

```diff
@@ -326,6 +393,218 @@ impl StacksBlockBuilder {
+                    let mut tx = burndb.tx_begin()?;
+                    let parent_stacks_block_snapshot = BurnDB::get_block_snapshot_for_winning_stacks_block(&mut tx, parent_burn_header_hash, parent_block_hash)?
+                        .ok_or(Error::NoSuchBlockError)?;
+
+                    parent_stacks_block_snapshot
+                };
+
+                let parent_chain_tip = StacksChainState::get_anchored_block_header_info(&chainstate.headers_db, parent_burn_header_hash, parent_block_hash)?
+                    .ok_or(Error::NoSuchBlockError)?;
+
+                let new_work = StacksWorkScore {
+                    burn: parent_stacks_block_snapshot.total_burn,
+                    work: parent_chain_tip.block_height.checked_add(1).expect("FATAL: block height overflow")
+                };
+
+                StacksBlockBuilder::from_parent_pubkey_hash(0, &parent_chain_tip, &new_work, &proof, pubkey_hash)
+            };
+
+        Ok(builder)
+    }
+
+    /// Given access to the mempool, mine an anchored block with no more than the given execution cost.
+    pub fn build_anchored_block(burndb: &mut BurnDB,
+                                chainstate: &mut StacksChainState,
+                                mempool: &MemPoolDB,
+                                parent_burn_header_hash: &BurnchainHeaderHash,
+                                parent_block_hash: &BlockHeaderHash,
+                                proof: VRFProof,
+                                pubkey_hash: Hash160,
+                                coinbase_tx: &StacksTransaction,
+                                execution_budget: ExecutionCost) -> Result<StacksBlock, Error> {
+
+        if let TransactionPayload::Coinbase(..) = coinbase_tx.payload {} else {
+            return Err(Error::MemPoolError("Not a coinbase transaction".to_string()));
+        }
+
+        let mut builder = StacksBlockBuilder::make_block_builder(burndb, chainstate, parent_burn_header_hash, parent_block_hash, proof, pubkey_hash)?;
+
+        builder.set_max_cost(execution_budget.clone());
+        let cost_overflow_recovery_builder = builder.clone();     // used to construct the real anchored block if we run out of budget
+        let mut header_reader_chainstate = chainstate.reopen()?;    // uesd for reading block headers during an epoch
+        
+        let mut tip_height = MemPoolDB::get_block_height(mempool.conn(), parent_burn_header_hash, parent_block_hash)?.unwrap_or(0);
+        let mut tip_burn_header_hash = parent_burn_header_hash.clone();
+        let mut tip_block_hash = parent_block_hash.clone();
+        
+        let mut epoch_tx = builder.epoch_begin(chainstate)?;
+        builder.try_mine_tx(&mut epoch_tx, coinbase_tx)?;
+
+        let mut considered = HashSet::new();
+        let mut next_timestamp = match MemPoolDB::get_next_timestamp(mempool.conn(), parent_burn_header_hash, parent_block_hash, 0)? {
+            Some(ts) => ts,
+            None => {
```

Comment:
In this scenario:

Block 1: receives 2 x block_limit transactions
Block 2: receives 0 transactions.

This check would mine an empty block at Block 2, right?

Reply:
I'm not sure how to parse `2 x block_limit`?

Reply:
If, in block 1, the mempool accepts a bunch of transactions, more than the block limit, then when it mines, there will be many transactions that weren't included, but could be included in the next block.

In block 2, if there are no new transactions admitted to the mempool, this check will return None, and cause the miner to mine an empty block before considering transactions that were accepted to the mempool earlier.

Reply:
Ah, yes, I see.  I'll fix that and add a test for it.  Block 2 should attempt to mine transactions from prior chain tips.

Reply:
This should be fixed now.  There's a test for it (`test_build_anchored_blocks_empty_chaintips`).